### PR TITLE
Close the event handles in the uv-loop in FileWatcher

### PR DESCRIPTION
**Pull request by CMurtagh-LGTM**
As libuv is a c library we need to call the constructor for the type explicitly.
Possibly fixes segfault #736
Importantly https://github.com/libuv/libuv/blob/v1.x/include/uv.h#L437 needs to be set so the polymorphism can work. One important instance of the polymorphism is calling `uv_close` with the reinterpret cast.

### CODE REVIEW

File: module/extension/FileWatcher/src/FileWatcher.cpp [hunk 1]
```diff
@@ -236,6 +236,7 @@ namespace module::extension {
                 if (paths.find(dir) == paths.end()) {
                     auto& p        = paths[path];
                     p.handle       = std::make_unique<uv_fs_event_t>();
+                    uv_fs_event_init(loop.get(), p.handle);
```
Comment (TrentHouliston) [inline module/extension/FileWatcher/src/FileWatcher.cpp:239]: Is there a delete that needs to be run that should be attached to the deleter?

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: module/extension/FileWatcher/src/FileWatcher.cpp (hunk 1 above)
Comment (CMurtagh-LGTM) [reply module/extension/FileWatcher/src/FileWatcher.cpp:239]: Its freed here https://github.com/NUbots/NUbots/blob/main/module/extension/FileWatcher/src/FileWatcher.cpp#L297

Comment (CMurtagh-LGTM) [reply module/extension/FileWatcher/src/FileWatcher.cpp:239]: I could add it to the deleter.

File: module/extension/FileWatcher/src/FileWatcher.hpp [hunk 2]
```diff
@@ -23,6 +23,20 @@
 
 namespace module::extension {
 
+
+    ///@brief Call the cleanup function and delete the ptr
+    ///@tparam must be be reinterpret castable into uv_handle_t
```
Comment (KipHamiltons) [inline module/extension/FileWatcher/src/FileWatcher.hpp:28]:
```suggestion
    /// @tparam must be be reinterpret castable into uv_handle_t
```
The one above as well. Always reviewing the important things 😂😂

**Review by KipHamiltons — COMMENTED** [review]
Stuff like this is impossible to review on my phone lol, but I did spot this 😝😝😝

File: module/extension/FileWatcher/src/FileWatcher.cpp [hunk 3]
```diff
@@ -110,9 +110,9 @@ namespace module::extension {
     FileWatcher::FileWatcher(std::unique_ptr<NUClear::Environment> environment)
         : Reactor(std::move(environment))
         , loop(std::make_unique<uv_loop_t>())
-        , add_watch(std::make_unique<uv_async_t>())
```
Comment (TrentHouliston) [inline module/extension/FileWatcher/src/FileWatcher.cpp:113]: Not a fan of new...
I think I'd rather either a custom wrapper struct for these, or shared_ptrs so you can set the deleter per object

File: module/extension/FileWatcher/src/FileWatcher.cpp [hunk 4]
```diff
@@ -291,22 +292,20 @@ namespace module::extension {
     }
 
     FileWatcher::~FileWatcher() {
-        for (const auto& path : paths) {
+        // Remove all the handles to the uv loop will finish
+        for (auto& path : paths) {
             if (path.second.handle) {
-                uv_fs_event_stop(path.second.handle.get());
-                uv_close(reinterpret_cast<uv_handle_t*>(path.second.handle.get()), [](uv_handle_t* /* handle */) {});
+                path.second.handle.reset();
```
Comment (TrentHouliston) [inline module/extension/FileWatcher/src/FileWatcher.cpp:298]: Do we need these anymore if we have the variables? they'll delete themselves anyway

Comment (CMurtagh-LGTM) [reply module/extension/FileWatcher/src/FileWatcher.cpp:298]: As per the comment, they need to be freed for libuv to finish. We wait for libuv to finish in the deconstructor of FileWatcher.

**Comment by CMurtagh-LGTM** [comment]
I've determined all the reworking of the Filewatcher is out of scope for this PR and that I'll just fix #736 here.

**Review by TrentHouliston — APPROVED** [review]
¯\_(ツ)_/¯

**Pull request merged**